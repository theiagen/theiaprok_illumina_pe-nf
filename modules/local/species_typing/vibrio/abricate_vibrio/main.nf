process ABRICATE_VIBRIO {
    tag "$meta.id"
    label 'process_single'

    container "us-docker.pkg.dev/general-theiagen/staphb/abricate:1.0.1-vibrio-cholera"

    input:
    tuple val(meta), path(assembly)
    val database_type 
    val min_percent_identity
    val min_percent_coverage

    output:
    tuple val(meta), path("*_abricate_hits.tsv"), emit: abricate_hits
    tuple val(meta), path("*_value.txt"), emit: abricate_value_results
    path("versions.yml"), emit: versions

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def database = database_type ?: "vibrio"
    def min_percent_cov = min_percent_coverage ? "--mincov ${min_percent_coverage}": ""
    def min_percent_ident = min_percent_identity ? "--minid ${min_percent_identity}": ""
    """
    date | tee DATE    
    # run abricate
    abricate \
      --db ${database} \
      ${min_percent_ident} \
      ${min_percent_cov} \
      --threads ${task.cpus} \
      --nopath \
      ${assembly} > ${prefix}_abricate_hits.tsv
    
    # presence or absence genes - ctxA, ompW and toxR
    # if empty report as (not detected)
    if grep -q '\\tctxA\\t' ${prefix}_abricate_hits.tsv; then echo "(not detected)" > CTXA.txt; else echo "present" > abricate_vibrio_CTXA_value.txt; fi
    if grep -q '\\tompW\\t' ${prefix}_abricate_hits.tsv; then echo "(not detected)" > OMPW.txt; else echo "present" > abricate_vibrio_OMPW_value.txt; fi
    if grep -q '\\ttoxR\\t' ${prefix}_abricate_hits.tsv; then echo "(not detected)" > TOXR.txt; else echo "present" > abricate_vibrio_TOXR_value.txt; fi

    # biotype - tcpA classical or tcpA ElTor
    if grep -q '\ttcpA_classical\t' ${prefix}_abricate_hits.tsv; then
      echo "tcpA_classical" > abricate_vibrio_BIOTYPE_value.txt
    elif grep -q '\ttcpA_ElTor\t' ${prefix}_abricate_hits.tsv; then
      echo "tcpA_ElTor" > abricate_vibrio_BIOTYPE_value.txt
    else
      echo "(not detected)" > abricate_vibrio_BIOTYPE_value.txt
    fi

    # serogroup - O1 or O139
    if grep -q '\twbeN_O1\t' ${prefix}_abricate_hits.tsv; then
      echo "O1" > abricate_vibrio_SEROGROUP_value.txt
    elif grep -q '\twbfR_O139\t' ${prefix}_abricate_hits.tsv; then
      echo "O139" > abricate_vibrio_SEROGROUP_value.txt
    else
      echo "(not detected)" > abricate_vibrio_SEROGROUP_value.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        abricate: \$(abricate -v)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_abricate_hits.tsv
    touch abricate_vibrio_CTXA.txt
    touch abricate_vibrio_OMPW.txt
    touch abricate_vibrio_TOXR.txt
    touch abricate_vibrio_BIOTYPE.txt
    touch abricate_vibrio_SEROGROUP.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        abricate: \$(abricate -v)
    END_VERSIONS
    """

}