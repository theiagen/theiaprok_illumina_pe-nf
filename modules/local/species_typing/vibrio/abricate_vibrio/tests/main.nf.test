nextflow_process {
    name "Test Process ABRICATE_VIBRIO"
    script "../main.nf"
    process "ABRICATE_VIBRIO"
    
    test("abricate_vibrio test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file("https://raw.githubusercontent.com/theiagen/test-datasets/refs/heads/vibrio/data/SRR7062531_filtered_contigs.fasta", checkIfExists: true)
                ])
                input[1] = "vibrio"  // database_type
                input[2] = 90        // min_percent_identity
                input[3] = 80        // min_percent_coverage
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.abricate_hits,
                    process.out.ctxa,
                    process.out.ompw,
                    process.out.toxr,
                    process.out.biotype,
                    process.out.serogroup,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("abricate_vibrio - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file(params.modules_testdata_base_path + "vibrio_test.fasta", checkIfExists: true)
                ])
                input[1] = "vibrio"  // database_type
                input[2] = 90        // min_percent_identity
                input[3] = 80        // min_percent_coverage
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
