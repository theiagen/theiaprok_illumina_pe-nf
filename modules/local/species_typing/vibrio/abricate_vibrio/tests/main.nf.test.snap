{
    "abricate_vibrio test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_abricate_hits.tsv:md5,a7fd457eb9d5487341ce7a57d3132d7e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "CTXA.txt:md5,e0db08a61c51d911c9e4fe3f610c6175"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "OMPW.txt:md5,e0db08a61c51d911c9e4fe3f610c6175"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "TOXR.txt:md5,e0db08a61c51d911c9e4fe3f610c6175"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "BIOTYPE.txt:md5,abe5a6b7c2c63600da6f663929fcfb87"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "SEROGROUP.txt:md5,b245400af5ccc23c31ff441a8f0628bb"
                ]
            ],
            [
                "versions.yml:md5,9bfa0dcf6d7d5f909910147e671b8489"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-26T14:04:31.752016347"
    },
    "abricate_vibrio - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_abricate_hits.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "CTXA.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "OMPW.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TOXR.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "BIOTYPE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SEROGROUP.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,9bfa0dcf6d7d5f909910147e671b8489"
                ],
                "abricate_hits": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_abricate_hits.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "biotype": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "BIOTYPE.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "ctxa": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "CTXA.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "ompw": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "OMPW.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "serogroup": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SEROGROUP.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "toxr": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TOXR.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9bfa0dcf6d7d5f909910147e671b8489"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-26T14:04:36.870991818"
    }
}